### PR TITLE
feat: add regression reliability metrics (EPE distribution, PICP, error-variance correlation)

### PR DESCRIPTION
**Pull request by Whatsonyourmind**
## What

Implements the regression-reliability metrics requested in #82 as a new `trustlens/metrics/regression.py` module (metrics-only — see Scope).

Three functions, matching the existing metric-module conventions (raw-array inputs, plain-dict returns, NumPy docstrings, graceful `status="skipped"` degradation):

- **`error_distribution(y_true, y_pred)`** — the EPE / absolute-error distribution: `median_absolute_error` (MedAE), `p90_absolute_error`, `max_error`, `mean_absolute_error`, `rmse`, plus histogram data for plotting.
- **`prediction_interval_coverage(y_true, lower, upper, confidence_level=0.95)`** — PICP. Returns coverage vs. nominal, `calibration_error`, `mean_interval_width`, and a verdict (over-/under-/well-calibrated). **Degrades gracefully** to `{"status": "skipped", ...}` when no intervals are supplied (sklearn regressors expose only point predictions).
- **`error_variance_correlation(y_true, y_pred, predicted_variance)`** — Pearson + Spearman between predicted uncertainty and actual `|error|` (does the model know where it's wrong?). Degrades gracefully when no variance is supplied; zero-variance inputs return `0.0` rather than `NaN`.

## Scope

Deliberately **metrics-module-only**, per `CONTRIBUTING.md` §5 ("`api.py` integration is optional Step 3") and the "one PR per feature" guideline. The issue's other DoD items — `analyze()` auto-dispatch on continuous targets, the visualization functions, and the `prediction_contract` / backend changes to admit regressors — are a larger cross-cutting change (they touch `api.py`, `core/pipeline.py`, `backends/sklearn.py`, `report.py`, `trust_score.py`) and are cleaner as follow-ups. This PR lands the correct, tested primitives; happy to follow up with the wire-in.

## Tests

`tests/test_regression.py` — **16 tests, all passing**: perfect-predictor and analytically-known MedAE/p90, `make_regression` integration (per the DoD), the graceful-skip paths, full/zero/normal-interval PICP, informative vs. uninformative variance correlation, constant-variance → 0, and shape-mismatch errors.

- `pytest tests/test_regression.py` → 16 passed
- existing `tests/test_metrics_all_exports.py` + `tests/test_api.py` still pass (no regressions)
- `ruff check`, `ruff format --check`, `mypy` all clean

Refs #82.

## Summary by CodeRabbit

* **New Features**
  * Added a regression diagnostics metrics module with three new metrics: error distribution, prediction interval coverage calibration, and error–uncertainty variance correlation.
  * Extended the metrics public API to re-export these new regression metrics for direct use.
* **Tests**
  * Added comprehensive tests covering correctness, edge cases, input validation, and export/public-contract checks for all regression metrics.
* **Documentation**
  * Updated the changelog with the new regression metrics under the Unreleased section.

### CODE REVIEW

**Comment by Whatsonyourmind** [comment]
Thanks for the review — addressed both inline points: a positive-integer guard on `n_bins` in `error_distribution`, and range guards on `confidence_level` ((0,1)) and `tolerance` ([0,1)) in `prediction_interval_coverage`, each with a test (now 18 passing; ruff/format/mypy clean).

On the docstring-coverage check: the ~27% is driven by the test methods in `tests/test_regression.py`, which I left undocumented to match the repo's existing test files (`test_failure.py` etc.) and the `ruff` config that ignores `D` rules for `tests/*`. All public functions and the private helpers carry full NumPy docstrings — happy to add per-test docstrings if you'd prefer.

**Review by Khanz9664 — COMMENTED** [review]
@Whatsonyourmind Thanks for the contribution and for keeping the scope focused.

I've completed a full review of the implementation, the associated tests, and the follow-up changes addressing the review feedback.

Overall, this is a well-scoped and well-executed first step toward the broader Regression Support roadmap tracked in #82.

A few things I particularly appreciated:

* The implementation follows the existing TrustLens metric-module conventions closely (raw array inputs, plain dictionary outputs, no unnecessary abstractions).
* `prediction_interval_coverage()` and `error_variance_correlation()` degrade gracefully when uncertainty information is unavailable, which is important because many common regressors only expose point predictions.
* Input validation is comprehensive, including shape checks, empty-input handling, interval validation, and the recently added parameter-range guards.
* Test coverage is thorough and includes both analytical cases and `make_regression` integration tests, along with a good set of edge cases.
* The documentation quality is strong and aligns well with the project's current documentation standards.

I also agree with the decision to keep this PR metrics-only. The remaining work required for full regression support (pipeline dispatch, backend integration, report integration, visualizations, and regression-specific reporting) is substantial and is better handled in follow-up contributions rather than expanding the scope of this PR.

One small note for future follow-up work: some returned plotting artifacts (such as histogram arrays) are currently NumPy arrays, which may eventually require consideration when the regression metrics are surfaced through report serialization and export workflows. I do not consider that a blocker for this PR.

From my review, the implementation successfully delivers the foundational regression metrics layer requested in #82 and does so in a maintainable way.

I'll proceed with workflow approval pending successful CI completion.

Thanks again for the contribution.

**Comment by Khanz9664** [comment]
@Whatsonyourmind The remaining failure appears to be a mypy typing issue in `trustlens/metrics/regression.py` rather than a functional problem with the implementation.

Current CI output:

```text
trustlens/metrics/regression.py:343: error: Need type annotation for "ranks"  [var-annotated]
trustlens/metrics/regression.py:350: error: Returning Any from function declared to return "ndarray[Any, dtype[Any]]"  [no-any-return]
```

Both errors appear to be inside `_average_ranks()`.

Could you please update the typing there and rerun the checks? Once the type-checking workflow is green, I'll do a final pass.

Everything else looks good from my review so far.

**Comment by Whatsonyourmind** [comment]
Fixed in `94c8d1a`. The two errors only surface under the CI invocation `mypy trustlens/ --follow-imports=skip` (without that flag they pass, which is why I missed them locally — reproduced it now). Both were in `_average_ranks()`:

- L343 `[var-annotated]` → annotated the buffer: `ranks: np.ndarray = np.empty(...)`.
- L350 `[no-any-return]` → returned through a typed local (`averaged: np.ndarray = avg[inv]; return averaged`) so the return is `ndarray`, not `Any`.

No behaviour change. Verified locally against the exact CI command: `mypy trustlens/ --follow-imports=skip` → **Success: no issues found in 35 source files**; ruff + ruff-format clean; `tests/test_regression.py` 18/18 pass. Should be green for your final pass — thanks!

**Review by Khanz9664 — APPROVED** [review]

**Comment by Khanz9664** [comment]
@Whatsonyourmind Thanks for the quick turnaround on the CI issue and for reproducing the exact mypy configuration used in the workflow.

I've reviewed the follow-up changes and confirmed that the remaining failures were limited to typing annotations within `_average_ranks()`. The fixes are minimal, maintain the existing behavior, and all checks are now passing.

This PR successfully delivers the foundational regression metrics layer requested in #82 while keeping the scope appropriately focused. The remaining regression-support work (pipeline integration, visualization, reporting, and trust scoring) will continue as follow-up efforts.

Thanks again for the contribution and for being responsive throughout the review process. 🚀

**Comment by Whatsonyourmind** [comment]
Thanks @Khanz9664 — really glad the foundational regression-metrics layer landed, and thanks for the careful review and merge. 🚀

Happy to keep going on the remaining #82 work. If it's useful, I'd suggest a focused next PR that wires these metrics into the evaluation/reporting path — so PICP, the EPE distribution, and the error-variance correlation surface in a regression report the way the classification metrics already do — then visualization as a separate scoped PR after, keeping each one small and reviewable like this one. Any preference on ordering, or an existing report/interface format you'd like me to match so it slots in cleanly?

**Comment by Khanz9664** [comment]
Thanks again for the contribution — the metrics layer was a solid first step and the implementation quality made the review process straightforward.

I agree with the proposed ordering. My preference would be:

1. **Regression pipeline & TrustReport integration**

   * Detect regression targets and route through a regression analysis path.
   * Surface the regression metrics in the report object.
   * Add regression report rendering (`report.show()`).
   * Ensure serialization support (`to_dict()`).

2. **Regression visualizations**

   * Residual analysis plot.
   * Error distribution visualization.
   * Any visualization-specific report integration.

3. **Regression trust score**

   * This will likely require a separate design discussion since the current trust score is classification-oriented and we'll need to think carefully about how uncertainty calibration and error behavior should contribute to a regression-specific score.

For consistency, I'd recommend following the existing classification architecture as closely as possible (metrics → report container → rendering layer), while keeping regression-specific logic isolated where appropriate.

If you'd like to take the next phase, feel free to open a focused PR for the pipeline/report integration work and reference #82. Keeping the work scoped into reviewable chunks has worked very well so far.

Thanks again for the continued interest in improving TrustLens. 🚀
